src/test/rustdoc-ui/check-doc-attrs.rs

```diff
@@ -0,0 +1,18 @@
+#![doc(inline)]
+//~^ WARNING can only be applied to a `use` item
+
+#[doc(inline)]
+#[doc(no_inline)]
+//~^^ ERROR conflicting doc inlining attributes
+//~|  ERROR conflicting doc inlining attributes
```

Comment:
Why is this duplicated?

Reply:
It warns one per attribute: the first one warns about `doc(inline)`, the second about `doc(no_inline)`.

Reply:
It's supposed to error only once though. I think it's just how the `MultiSpan` is handled.

Reply:
I added a `#[doc(no_inline)]` and I'm getting (with `-Zdeduplicate-diagnostics=no`):

```
error: conflicting doc inlining attributes
 --> src/test/rustdoc-ui/check-doc-attrs.rs:4:7
  |
4 | #[doc(inline)]
  |       ^^^^^^ this attribute...
5 | #[doc(no_inline)]
  |       ^^^^^^^^^ ...conflicts with this attribute
  |
  = help: remove one of the conflicting attributes

error: conflicting doc inlining attributes
 --> src/test/rustdoc-ui/check-doc-attrs.rs:4:7
  |
4 | #[doc(inline)]
  |       ^^^^^^ this attribute...
5 | #[doc(no_inline)]
6 | #[doc(no_inline)]
  |       ^^^^^^^^^ ...conflicts with this attribute
  |
  = help: remove one of the conflicting attributes

error: conflicting doc inlining attributes
 --> src/test/rustdoc-ui/check-doc-attrs.rs:4:7
  |
4 | #[doc(inline)]
  |       ^^^^^^ this attribute...
5 | #[doc(no_inline)]
  |       ^^^^^^^^^ ...conflicts with this attribute
  |
  = help: remove one of the conflicting attributes

error: conflicting doc inlining attributes
 --> src/test/rustdoc-ui/check-doc-attrs.rs:4:7
  |
4 | #[doc(inline)]
  |       ^^^^^^ this attribute...
5 | #[doc(no_inline)]
6 | #[doc(no_inline)]
  |       ^^^^^^^^^ ...conflicts with this attribute
  |
  = help: remove one of the conflicting attributes
```

Reply:
Looks like two different `Use` items exist:
```
DEBUG rustdoc::visit_ast visiting item Item { ident: Bar#0, hir_id: HirId { owner: DefId(0:3 ~ check_doc_attrs[8787]::{misc#1}), local_id: 0 }, attrs: [Attribute { kind: Normal(AttrItem { path: Path { span: src/test/rustdoc-ui/check-doc-attrs.rs:4:3: 4:6 (#0), segments: [PathSegment { ident: doc#0, id: NodeId(12), args: None }], tokens: None }, args: Delimited(DelimSpan { open: src/test/rustdoc-ui/check-doc-attrs.rs:4:6: 4:7 (#0), close: src/test/rustdoc-ui/check-doc-attrs.rs:4:13: 4:14 (#0) }, Parenthesis, TokenStream([(Token(Token { kind: Ident("inline", false), span: src/test/rustdoc-ui/check-doc-attrs.rs:4:7: 4:13 (#0) }), Alone)])), tokens: None }, None), id: AttrId(1), style: Outer, span: src/test/rustdoc-ui/check-doc-attrs.rs:4:1: 4:15 (#0) }, Attribute { kind: Normal(AttrItem { path: Path { span: src/test/rustdoc-ui/check-doc-attrs.rs:5:3: 5:6 (#0), segments: [PathSegment { ident: doc#0, id: NodeId(13), args: None }], tokens: None }, args: Delimited(DelimSpan { open: src/test/rustdoc-ui/check-doc-attrs.rs:5:6: 5:7 (#0), close: src/test/rustdoc-ui/check-doc-attrs.rs:5:16: 5:17 (#0) }, Parenthesis, TokenStream([(Token(Token { kind: Ident("no_inline", false), span: src/test/rustdoc-ui/check-doc-attrs.rs:5:7: 5:16 (#0) }), Alone)])), tokens: None }, None), id: AttrId(2), style: Outer, span: src/test/rustdoc-ui/check-doc-attrs.rs:5:1: 5:18 (#0) }], kind: Use(Path { span: src/test/rustdoc-ui/check-doc-attrs.rs:8:9: 8:17 (#0), res: Def(Struct, DefId(0:5 ~ check_doc_attrs[8787]::bar::Bar)), segments: [PathSegment { ident: bar#0, hir_id: Some(HirId { owner: DefId(0:3 ~ check_doc_attrs[8787]::{misc#1}), local_id: 1 }), res: Some(Def(Mod, DefId(0:4 ~ check_doc_attrs[8787]::bar))), args: None, infer_args: false }, PathSegment { ident: Bar#0, hir_id: Some(HirId { owner: DefId(0:3 ~ check_doc_attrs[8787]::{misc#1}), local_id: 2 }), res: Some(Err), args: None, infer_args: false }] }, Single), vis: Spanned { node: Public, span: src/test/rustdoc-ui/check-doc-attrs.rs:8:1: 8:4 (#0) }, span: src/test/rustdoc-ui/check-doc-attrs.rs:8:1: 8:18 (#0) }
```
```
DEBUG rustdoc::visit_ast visiting item Item { ident: Bar#0, hir_id: HirId { owner: DefId(0:8 ~ check_doc_attrs[8787]::{misc#1}::{misc#0}), local_id: 0 }, attrs: [Attribute { kind: Normal(AttrItem { path: Path { span: src/test/rustdoc-ui/check-doc-attrs.rs:4:3: 4:6 (#0), segments: [PathSegment { ident: doc#0, id: NodeId(12), args: None }], tokens: None }, args: Delimited(DelimSpan { open: src/test/rustdoc-ui/check-doc-attrs.rs:4:6: 4:7 (#0), close: src/test/rustdoc-ui/check-doc-attrs.rs:4:13: 4:14 (#0) }, Parenthesis, TokenStream([(Token(Token { kind: Ident("inline", false), span: src/test/rustdoc-ui/check-doc-attrs.rs:4:7: 4:13 (#0) }), Alone)])), tokens: None }, None), id: AttrId(1), style: Outer, span: src/test/rustdoc-ui/check-doc-attrs.rs:4:1: 4:15 (#0) }, Attribute { kind: Normal(AttrItem { path: Path { span: src/test/rustdoc-ui/check-doc-attrs.rs:5:3: 5:6 (#0), segments: [PathSegment { ident: doc#0, id: NodeId(13), args: None }], tokens: None }, args: Delimited(DelimSpan { open: src/test/rustdoc-ui/check-doc-attrs.rs:5:6: 5:7 (#0), close: src/test/rustdoc-ui/check-doc-attrs.rs:5:16: 5:17 (#0) }, Parenthesis, TokenStream([(Token(Token { kind: Ident("no_inline", false), span: src/test/rustdoc-ui/check-doc-attrs.rs:5:7: 5:16 (#0) }), Alone)])), tokens: None }, None), id: AttrId(2), style: Outer, span: src/test/rustdoc-ui/check-doc-attrs.rs:5:1: 5:18 (#0) }], kind: Use(Path { span: src/test/rustdoc-ui/check-doc-attrs.rs:8:9: 8:17 (#0), res: Def(Ctor(Struct, Const), DefId(0:6 ~ check_doc_attrs[8787]::bar::Bar::{constructor#0})), segments: [PathSegment { ident: bar#0, hir_id: Some(HirId { owner: DefId(0:8 ~ check_doc_attrs[8787]::{misc#1}::{misc#0}), local_id: 1 }), res: Some(Err), args: None, infer_args: false }, PathSegment { ident: Bar#0, hir_id: Some(HirId { owner: DefId(0:8 ~ check_doc_attrs[8787]::{misc#1}::{misc#0}), local_id: 2 }), res: Some(Err), args: None, infer_args: false }] }, Single), vis: Spanned { node: Public, span: src/test/rustdoc-ui/check-doc-attrs.rs:8:1: 8:4 (#0) }, span: src/test/rustdoc-ui/check-doc-attrs.rs:8:9: 8:17 (#0) }
```

Both point to line 8 (`pub use bar::Bar`). Seems very confusing to me.

I think this could be cleaned up later, considering that the error get deduplicated by rustc anyway.

Reply:
I think that if `check_item_attrs` gets called twice, we should fix it first before merging this one. The problem is that you call the check on the `krate` and then it gets called again. Instead of doing that, just check if you are at the crate level. I did the same in another part of the compiler, posting the link when I have it...

Reply:
Hum but in your case, just removing the function call should be enough to remove this issue. :3

Reply:
Sorry, I still don't understand... Which call would you remove? And from my debugging it appears to me that there are two distinct `Item`s pointing to the same `use` statement, is this wrong?

Reply:
I really think the common item is `krate.item`.

Reply:
The issue is that `struct S;` desugars to something like
```rust
struct S {}
fn S() -> S { ... }
```
except with compiler magic so you don't need the parentheses. If you look at the debugging @LeSeulArtichaut found, it contains
```
kind: Use(Path { span: src/test/rustdoc-ui/check-doc-attrs.rs:8:9: 8:17 (#0), res: Def(Struct, DefId(0:5 ~ check_doc_attrs[8787]::bar::Bar))
kind: Use(Path { span: src/test/rustdoc-ui/check-doc-attrs.rs:8:9: 8:17 (#0), res: Def(Ctor(Struct, Const), DefId(0:6 ~ check_doc_attrs[8787]::bar::Bar::{constructor#0}))
```
So I think the fix is to ignore attributes on Ctor items, because they can never occur without also occurring on the struct itself.

Reply:
For the same reason, re-exporting an item have different behavior than using the original if it has private fields: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=764be855d78364df0b1b8eb0dbd7ce8e